### PR TITLE
Adding the logic of in-warning regexes

### PR DESCRIPTION
**Pull request by akuleshov7**
### What's done:
- Regex support

### CODE REVIEW

File: save-plugins/warn-plugin/src/commonNonJsTest/kotlin/org/cqfn/save/plugin/warn/utils/ComparisonTest.kt [hunk 1]
```diff
@@ -1,18 +1,13 @@
...
-import org.cqfn.save.plugin.warn.WarnPlugin
 import org.cqfn.save.plugin.warn.WarnPluginConfig
 
 import okio.FileSystem
 
 import kotlin.test.AfterTest
 import kotlin.test.BeforeTest
 import kotlin.test.Test
-import kotlin.test.assertEquals
-import kotlin.test.assertTrue
```
Comment (kgevorkyan) [inline save-plugins/warn-plugin/src/commonNonJsTest/kotlin/org/cqfn/save/plugin/warn/utils/ComparisonTest.kt:15]: Why there is no `unresolved reference` err after these changes?

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        val openingDelimiter = warnPluginConfig.patternForRegexInWarning!![0]
+        val closingDelimiter = warnPluginConfig.patternForRegexInWarning[1]
+
+        // searching all delimited regex in the warning
+        val foundSubStringsWithRegex = this
+            .findDelimitedSubStringsWith(openingDelimiter, closingDelimiter)
+            .entries
+            .sortedBy { it.key }
+
+        val resultWithRegex = foundSubStringsWithRegex.mapIndexed { i, entry ->
```
Comment (petertrr) [inline save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt:95]: Is this logic equivalent to something like
```kotlin
foundSubStringsWithRegex.foldRight(this) { acc, entry ->
            acc.removeRange(entry.value, entry.value + closingDelimiter.length)
                .removeRange(entry.key, entry.key + closingDelimiter.length)
}
```
?

**Review by petertrr — COMMENTED** [review]
Should we describe this format in warn-plugin's readme?

UPD: done [Andrey Kuleshov]

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPluginConfig.kt [hunk 3]
```diff
@@ -124,9 +131,12 @@ data class WarnPluginConfig(
         val newLineCaptureGroupOut = if (newWarningTextHasLine) (lineCaptureGroupOut ?: 2) else null
         val newColumnCaptureGroupOut = if (newWarningTextHasColumn) (columnCaptureGroupOut ?: 3) else null
         val newMessageCaptureGroupOut = messageCaptureGroupOut ?: 4
+
+        // FixMe: why these requires are at the end of logic? Should be moved to the beginning of the method? kgevorkyan
```
[the remark this answers is not in the document]
Comment (akuleshov7) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPluginConfig.kt:135]: Could you please create an issue for it? @kgevorkyan and fix it

File: save-plugins/warn-plugin/src/commonNonJsTest/kotlin/org/cqfn/save/plugin/warn/utils/ComparisonTest.kt (hunk 1 above)
Comment (akuleshov7) [reply save-plugins/warn-plugin/src/commonNonJsTest/kotlin/org/cqfn/save/plugin/warn/utils/ComparisonTest.kt]: yes, bug

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPluginConfig.kt [hunk 4]
```diff
@@ -145,25 +155,39 @@ data class WarnPluginConfig(
...
+            patternForRegexInWarning ?: defaultPatternForRegexInWarning
         )
     }
 
     private fun requirePositiveIfNotNull(value: Long?) {
         value?.let {
             require(value >= 0) {
                 """
-                    Error: All integer values in [warn] section of `$configLocation` config should be positive!
+                    [Critical Error]: All integer values in [warn] section of `$configLocation` config should be positive!
```
[the remark this answers is not in the document]
Comment (akuleshov7) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPluginConfig.kt:166]: agree

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPluginConfig.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (kgevorkyan) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPluginConfig.kt:135]: https://github.com/cqfn/save/issues/269

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt (hunk 2 above)
Comment (akuleshov7) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt:95]: discussed offline, the main aim of my logic is to escape special symbols

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    private fun createFailFromSingleMiss(baseText: String, warnings: List<Warning>) =
+            Fail("$baseText: $warnings", "$baseText (${warnings.size})")
+
+    private fun createFailFromDoubleMiss(missingWarnings: List<Warning>, unexpectedWarnings: List<Warning>) =
+            Fail(
+                "$EXPECTED_BUT_NOT_RECEIVED: $missingWarnings, and ${UNEXPECTED.lowercase()}: $unexpectedWarnings",
+                "$EXPECTED_BUT_NOT_RECEIVED (${missingWarnings.size}), and ${UNEXPECTED.lowercase()} (${unexpectedWarnings.size})"
+            )
+
+    private fun String.createRegexFromMessage(): Regex {
```
[the remark this answers is not in the document]
Comment (akuleshov7) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt:84]: Yes, sorry for this, added

File: save-common/src/commonMain/kotlin/org/cqfn/save/core/plugin/PluginConfig.kt [hunk 6]
```diff
@@ -128,7 +128,7 @@ data class GeneralConfig(
          * Default regex for expected warnings in test resources, e.g.
          * `// ;warn:2:4: Class name in incorrect case`
          */
-        val defaultInputPattern = Regex("// ;warn:(.+):(\\d+): (.+)")
+        val defaultExpectedWarningPattern = Regex("// ;warn:?(.*):(\\\\d*): (.+)")
```
[the remark this answers is not in the document]
Comment (akuleshov7) [reply save-common/src/commonMain/kotlin/org/cqfn/save/core/plugin/PluginConfig.kt]: yes, previously worked correctly. Two slashes  - is my bug

**Review by petertrr — APPROVED** [review]

**Pull request merged**